packages/runtime/container-runtime/src/scheduleManager.ts

```diff
@@ -0,0 +1,310 @@
+import EventEmitter from "events";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert, IsoBuffer, performance } from "@fluidframework/common-utils";
+import { IDeltaManager } from "@fluidframework/container-definitions";
+import { DataCorruptionError, extractSafePropertiesFromMessage } from "@fluidframework/container-utils";
+import { isUnpackedRuntimeMessage } from "@fluidframework/driver-utils";
+import { ISequencedDocumentMessage, IDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ChildLogger } from "@fluidframework/telemetry-utils";
+import { compress } from "lz4js";
+import { latencyThreshold } from "./connectionTelemetry";
+import { DeltaScheduler } from "./deltaScheduler";
+import { pkgVersion } from "./packageVersion";
+
+type IRuntimeMessageMetadata = undefined | {
+    batch?: boolean;
+};
+
+/**
+ * This class controls pausing and resuming of inbound queue to ensure that we never
+ * start processing ops in a batch IF we do not have all ops in the batch.
+ */
+ class ScheduleManagerCore {
+    private pauseSequenceNumber: number | undefined;
+    private currentBatchClientId: string | undefined;
+    private localPaused = false;
+    private timePaused = 0;
+    private batchCount = 0;
+
+    constructor(
+        private readonly deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>,
+        private readonly logger: ITelemetryLogger,
+    ) {
+        // Listen for delta manager sends and add batch metadata to messages
+        this.deltaManager.on("prepareSend", (messages: IDocumentMessage[]) => {
+            if (messages.length === 0) {
+                return;
+            }
+
+            // First message will have the batch flag set to true if doing a batched send
+            const firstMessageMetadata = messages[0].metadata as IRuntimeMessageMetadata;
+            if (!firstMessageMetadata?.batch) {
+                return;
+            }
+
+            // If the batch contains only a single op, clear the batch flag.
+            if (messages.length === 1) {
+                delete firstMessageMetadata.batch;
+                return;
+            }
+
+            messages.map((msg) => {
+                const parsedMessage = JSON.parse(msg.contents);
+                const innerContents = JSON.stringify(parsedMessage.contents);
```

Comment:
I'm not sure why you are extracting contents.contents out of message.
if that's required, it's clearly very expensive work to parse & stringify back. While you are not adding here (but rahter leveraging existing ugly way we deal with batches), I'd suggest to shift that logic to submitRuntimeMessage().
This will require some staging, as it will require moving stringification / parsing of content from Loader layer (Container, DeltaManager classes) to runtime layer. That way, Loader layer just operates with payload as is (string | binary), and all the logic about how to package / compress data is in runtime.

It will be challenging for same reasons I mentioned above (needs to be somehow conditional), but I'm sure we can find a way to pull it off. Let's chat about it

Reply:
@vladsud  do you mind adding me to the chat?  I'm curious to understand it better. Taking in consideration the staging costs/risks + the cost of compressing every op  (up to 2x? compression gain), I wonder if it wouldn't be better to compress everything (5x compression gain) + encode base64 and work with the ODSP contacts to decode and uncompress on their end...

Reply:
@NicholasCouri, I'm not sure what problem you are trying to solve with ODSP uncompressing ops. Storage should be free to decompress, but it should not be required. And PUSH should not touch op content - compressed content should go in & out.

Reply:
I think I got it now but I was questioning the possibility/feasibility of switching the approach to compressing the entire batch instead of just the content of each op (what you called on the Batch document: **solid compression for op batches**. ) 

Justus was letting me know the plan is to eventually do that (Ex. Batch with 20 ops would become a batch with 1 compressed op (including all original 20) + 19 noops). 

I was also curious around the possibility of doing an approach to add steps 3,5,8 to the current Ops flow - I understand we probably want to keep the complexity on the client + avoid COGs implications but wanted to hear your thoughts around it.

1. Client submits compressed op batches (probably need to change the protocol from SubmitOps to add content type (txt,zip, etc.)
2. WebSocket server receives the compressed batch
3. **WebSocket server decompresses the batch**
4. WebSocket server constructs a Kafka message that includes the batch
5. **WebSocket server compresses that Kafka message**
6. WebSocket server sends that message to Kafka
7. Ordering service (Deli) consumes the Kafka message
8. **Deli decompresses the message**
9. Deli processes the batch and sequences the ops
10. Deli sends the sequenced messages to Kafka -> other backend servers consume the sequenced messages (no compression)

Reply:
I do not see the point for ordering service caring about compression. Plus, we want compressed payloads to low through all of the ordering service - it will make it substantially cheaper and scale better. Plus, fan-out (broadcasting ops to all clients) is the most expensive part, it should work with compressed content for efficiency.

The only place where we might want to always decompress, is storage, mostly to reduce back compat.